src/neuron/container/soa_container.hpp

```diff
@@ -171,6 +171,13 @@ void check_permutation_vector(Rng const& range, std::size_t size) {
 
 enum struct may_cause_reallocation { Yes, No };
 
+/** @brief Defer deleting pointers to deallocated memory.
+ *
+ *  The address of a pointer to the underlying storage of `field_data` can
+ *  escape the container. When deallocating the container the memory is
+ *  deallocated but the pointer to the storage location is set to `nullptr` and
+ *  "leaked" into this vector. These pointers are all `nullptr`s.
```

Comment:
Looking at usage of the `defer_delete(...)` method, I'm not sure that the `nullptr` part is actually true at the moment. I think that's a bug though, which could be fixed in the destructors of the two specialisations of the `field_data` struct.

I was also a bit tripped up by the formulation here. `void* foo = (*defer_delete_storage)[i];` is not `nullptr` for any `i`, but `foo` is the address of an array of `n` pointers (allocated on the heap), and each of those `n` pointers could/should be `nullptr`.

Reply:
> I'm not sure that the nullptr part is actually true at the moment.
Wouldn't this imply that there's dangling pointers in the wild? I thought the point of allocating the pointer on the heap was that we can give other parts of the code the address of this pointer and then set it to `nullptr` then the storage it normally points is deallocated. In doing so the code that has the address of the pointer knows if the storage is still valid (non `NULL`) vs. invalid (`NULL`).

Reply:
I **think** that it's currently just about OK, but that more consistently setting some pointers to `nullptr` would reduce the chances of errors creeping in in the future.

When a data handle is dereferenced then the first check is on whether it refers to a valid row in the global SoA storage (i.e. whether the referred-to value has been deleted). If the row is valid, we use it as an offset from the data pointer (which is the potentially-not-always-null-when-it-should-be thing under discussion here).

I think that means that to trigger a problem you would need the handle to point to a valid row, but for the `data()` pointer of the column to be invalid. I believe the only way you can have a valid row with a no-longer-valid column in the "real" NEURON application is if an optional column becomes disabled at runtime...and in that case, the `data()` **is** set to `nullptr` so all is fine.

Methods like `neuron::Model::delete_mechanism(...)` will refuse to proceed if the size is nonzero, so I think in non-contrived cases it's hard to trigger destruction of a non-empty `soa` structure (thus creating "valid" rows with invalid columns) -- but we could probably generate some ASan errors in unit tests that do that, and fix them by more pedantically setting `nullptr` in a few places.

Reply:
I've changed the wording to reflect the current state.